jena-text/src/main/java/org/apache/jena/query/text/DatasetGraphText.java

```diff
@@ -166,12 +166,33 @@ public void begin(TxnType txnType) {
 
     @Override
     public void begin(ReadWrite readWrite) {
-        // The "super.begin" is enough.
+        tdb2ResetCheck();
         readWriteMode.set(readWrite);
         super.begin(readWrite) ;
         super.getMonitor().start() ;
     }
 
+    // if a TDB2 compaction happened, the coordinator is a new object.
+    private void tdb2ResetCheck() {
+        if ( tdb2TransactionCoordinator != null ) {
+            TransactionCoordinator currentCoord = org.apache.jena.tdb2.sys.TDBInternal.getTransactionCoordinator(super.get());
+            if ( tdb2TransactionCoordinator != currentCoord ) {
+                tdb2Setup(currentCoord);
+                tdb2TransactionCoordinator = currentCoord;
+            }
+        }
+    }
+
+    private void tdb2Setup(TransactionCoordinator coord) {
+        // Does not overlap with the ids used by TDB2.
+        byte[] componentID = { 2,4,6,10 } ;
+        TransactionalComponent tc = new TextIndexDB(ComponentId.create(null, componentID), textIndex);
+        coord.modifyConfig(()->coord.add(tc));
+        commitAction = delegateCommit;
+        abortAction = delegateAbort;
+
+    }
```

Comment:
From the PR description, I think the fix is mainly in this part. I don't understand everything involved here, but couldn't find anything that looked wrong. +1 :+1:  Thanks!

Reply:
Yes - small fix, more changes in tests. Then some reshuffle (something bad when on with formatting at some point) - broken into separate commits.